impl/src/main/java/org/ehcache/impl/serialization/CompactJavaSerializer.java

```diff
@@ -130,57 +133,63 @@ public boolean equals(T object, ByteBuffer binary) throws ClassNotFoundException
     return object.equals(read(binary));
   }
 
-  private int getOrAddMapping(ObjectStreamClass desc) throws IOException {
+  private int getOrAddMapping(ObjectStreamClass desc) {
     SerializableDataKey probe = new SerializableDataKey(desc, false);
     Integer rep = writeLookup.get(probe);
-    if (rep != null) {
-      return rep;
-    }
-
-    // Install new rep - locking
-    lock.lock();
-    try {
+    if (rep == null) {
       return addMappingUnderLock(desc, probe);
-    } finally {
-      lock.unlock();
+    } else {
+      return rep;
     }
   }
 
   private int addMappingUnderLock(ObjectStreamClass desc, SerializableDataKey probe) {
-    ObjectStreamClass disconnected = disconnect(desc);
-    SerializableDataKey key = new SerializableDataKey(disconnected, true);
-    while (true) {
-      Integer rep = writeLookup.get(probe);
-      if (rep != null) {
-        return rep;
+    lock.lock();
+    try {
+      if (potentiallyInconsistent) {
+        refreshMappingsFromServer(readLookup.entrySet());
+        potentiallyInconsistent = false;
       }
-      rep = nextStreamIndex++;
-
-      ObjectStreamClass existingOsc = readLookup.putIfAbsent(rep, disconnected);
-      if (existingOsc == null) {
-        writeLookup.put(key, rep);
-        readLookupLocalCache.put(rep, disconnected);
-        return rep;
-      } else {
-        ObjectStreamClass discOsc = disconnect(existingOsc);
-        writeLookup.put(new SerializableDataKey(discOsc, true), rep);
-        readLookupLocalCache.put(rep, discOsc);
+      ObjectStreamClass disconnected = disconnect(desc);
+      SerializableDataKey key = new SerializableDataKey(disconnected, true);
+      while (true) {
+        Integer rep = writeLookup.get(probe);
+        if (rep != null) {
+          return rep;
+        }
+        rep = nextStreamIndex++;
```

Comment:
The single-step probing used in this method for updating the local association cache gives me pause.  

I can envision situations (for example, some other client has added a batch of complex cache objects that are composed of Java objects new to the cache) where many failed puts would be required before the local copy of the class associations is synchronized with the server and `nextStreamIndex` accurately reflects the server state.

Reply:
Discussed ... @chrisdennis is opening  a new issue to address this (needs a new server operation).